crates/storage/src/address.rs

```diff
@@ -0,0 +1,480 @@
+    #[must_use]
+    pub fn new() -> Self {
+        Self(Uuid::new_v4())
+    }
+
+    /// Returns a slice of 16 octets containing the value.
+    #[must_use]
+    pub const fn as_bytes(&self) -> &Bytes {
+        self.0.as_bytes()
+    }
+}
+
+impl Default for Id {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl Display for Id {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+impl From<Id> for Uuid {
+    fn from(id: Id) -> Self {
+        id.0
+    }
+}
+
+/// Path to an [`Element`](crate::entities::Element).
+///
+/// [`Element`](crate::entities::Element)s are stored in a hierarchical
+/// structure, and their path represents their location within that structure.
+/// Path segments are separated by a double-colon `::`, and all paths are
+/// absolute, and should start with a leading separator to enforce the clarity
+/// of this plus allow for future expansion of functionality.
+///
+/// [`Path`]s are case-sensitive, support Unicode, and are limited to 255
+/// characters in length. Note, the separators do NOT count towards this limit.
+///
+/// [`Path`]s are not allowed to be empty.
+///
+/// There is no formal limit to the levels of hierarchy allowed, but in practice
+/// this is limited to 255 levels (assuming a single byte per segment name).
+///
+#[derive(Clone, Debug, Eq, Hash, PartialEq, PartialOrd, Ord)]
+pub struct Path {
+    /// A list of path segment offsets, where offset `0` is assumed, and not
+    /// stored.
+    offsets: Vec<u8>,
+
+    /// The path to the element. This is a string of up to 255 characters in
+    /// length, and is case-sensitive. Internally the segments are stored
+    /// without the separators.
+    path: Flexstr<255>,
+}
+
+impl Path {
+    /// Creates a new [`Path`] from a string.
+    ///
+    /// # Parameters
+    ///
+    /// * `path` - The path to the [`Element`](crate::entities::Element).
+    ///
+    /// # Errors
+    ///
+    /// An error will be returned if:
+    ///
+    ///   - The path is empty, including if it contains only separators.
+    ///   - The path is too long (the maximum length allowed is 255 characters).
+    ///   - Any of the path segments are empty.
+    ///   - The path is not absolute, i.e. it does not start with a double colon
+    ///     separator `::`.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use calimero_storage::address::Path;
+    /// Path::new("::root::node::leaf").unwrap();
+    /// ```
+    ///
+    pub fn new<S: Into<String>>(path: S) -> Result<Self, PathError> {
+        let string = path.into();
+
+        if string.is_empty() {
+            return Err(PathError::Empty);
+        }
+        if !string.starts_with("::") {
+            return Err(PathError::NotAbsolute);
+        }
+
+        #[allow(clippy::string_slice)] // We know the string starts with `::`
+        let segments = string[2..].split("::").collect::<Vec<&str>>();
+
+        if segments.is_empty() {
```

Comment:
That's a really good suggestion - peeking to see if there's a next element at the beginning instead of allocating is more efficient. Thanks! 🙂

Reply:
I've ended up leaving this as it was - although peeking would save an allocation, it doesn't actually help because we don't know the length, and so we end up with a worse allocation of the `Vec` later, or a double iteration to find the length. That speed trade-off is worse than the minor extra storage of the last child being removed, and so I think it's fine as it is.

Reply:
```rust
let segments = string[2..].split("::").collect::<Vec<&str>>();
```

you only know the length because of the first allocation, at which point you didn't know the length, so you still have the "worse allocation of the `Vec`"

so you're still incurring the speed depreciation, and still having extra storage

worth noting that a `Vec<&str>` will also consume 2x more memory than a `Vec<u8>`, though it's unnecessary here

Reply:
I am not sure of the comparison to a `Vec<u8>`, as that's not something that is comparable here?

The best way to do this would be to manually loop over the string and work with slices, but I'm not interested in making micro-optimisations at this time... it is a distraction, and in an area that is insignificant. Hence I consider this "good enough", and certainly not bad.

Reply:
> I am not sure of the comparison to a Vec<u8>, as that's not something that is comparable here?

you need a `Vec<u8>`, which stores the offsets, but the current approach temporarily allocates a `Vec<&str>` to do that, which from what's been highlighted is unnecessary since the allocation of `Vec<&str>` incurs the same cost you mention you wanted to avoid

I don't think it's a micro-optimization, far from it, but certainly a non-blocker, approved